libs/x-content/src/main/java/org/elasticsearch/common/xcontent/MediaTypeParser.java

```diff
@@ -120,4 +125,37 @@ public T getMediaType() {
             return parameters;
         }
     }
+
+    public static class Builder<T extends MediaType> {
+        private final Map<String, T> formatToMediaType = new HashMap<>();
+        private final Map<String, T> typeWithSubtypeToMediaType = new HashMap<>();
+        private final Map<String, Map<String, Pattern>> parametersMap = new HashMap<>();
+
+        public Builder<T> withMediaTypeAndParams(String alternativeMediaType, T mediaType, Map<String, Pattern> paramNameAndValueRegex) {
+            typeWithSubtypeToMediaType.put(alternativeMediaType.toLowerCase(Locale.ROOT), mediaType);
+            formatToMediaType.put(mediaType.format(), mediaType);
+
+            Map<String, Pattern> parametersForMediaType = new HashMap<>(paramNameAndValueRegex.size());
+            for (Map.Entry<String, Pattern> params : paramNameAndValueRegex.entrySet()) {
+                String parameterName = params.getKey().toLowerCase(Locale.ROOT);
+                Pattern parameterRegex = params.getValue();
+                Pattern pattern = Pattern.compile(parameterRegex.pattern(), Pattern.CASE_INSENSITIVE);
```

Comment:
can you remove the insensitive recompile here and and just use the pattern as-is (and move the insenstivie up to the declration)

Reply:
I will apply `toLowerCase` for keys though

Reply:
OK ... in that case, maybe go back to a `Map<String, String>` as the collection passed ?  It is a bit confusing to compile it, then re-compile it.  Sorry, when i first suggested to pass `Map<String, Pattern>` I had envisioned only compiling once.

Reply:
agree, will refactor this back